rpc.go

```diff
@@ -176,56 +183,62 @@ func (t *Topic) Publish(
 
 	var respCh chan internalResponse
 	var msgID cid.Cid
+	msgID = cid.NewCidV1(cid.Raw, util.Hash(data))
 	if !args.ignoreResponse {
-		msgID = cid.NewCidV1(cid.Raw, util.Hash(data))
 		respCh = make(chan internalResponse)
-		t.lk.Lock()
-		t.resChs[msgID] = respCh
-		t.lk.Unlock()
 	}
+	t.lk.Lock()
+	t.ongoing[msgID] = ongoingMessage{
+		ctx:    ctx,
+		data:   data,
+		opts:   args.pubOpts,
+		respCh: respCh,
+	}
+	t.lk.Unlock()
 
 	if err := t.t.Publish(ctx, data, args.pubOpts...); err != nil {
-		return nil, fmt.Errorf("publishing to main topic: %v", err)
+		return nil, fmt.Errorf("publishing to topic: %v", err)
 	}
 
 	resultCh := make(chan Response)
-	if respCh != nil {
-		go func() {
-			defer func() {
-				t.lk.Lock()
-				delete(t.resChs, msgID)
-				t.lk.Unlock()
-				close(resultCh)
-			}()
-			for {
-				select {
-				case <-ctx.Done():
-					if !args.multiResponse {
-						resultCh <- Response{
-							Err: ErrResponseNotReceived,
-						}
-					}
+	if args.ignoreResponse {
+		close(resultCh)
+	}
+	go func() {
```

Comment:
Here we starts the goroutine regardless of `ignoreResponse` as we want to republish to new peers until deadline is hit